core/src/sumeragi/network_topology.rs

```diff
@@ -125,29 +128,125 @@ impl Topology {
         let rotate_at = self.min_votes_for_commit().min(self.sorted_peers.len());
-        self.sorted_peers[..rotate_at].rotate_left(1);
+        if rotate_at > 0 {
+            self.sorted_peers[..rotate_at].rotate_left(1);
+        }
     }
+
     /// Pull peers up in the topology to the top of the a set while preserving local order.
     pub fn lift_up_peers(&mut self, to_lift_up: &[PublicKey]) {
         self.sorted_peers
             .sort_by_cached_key(|peer| !to_lift_up.contains(&peer.public_key));
     }
+
+    /// Perform sequence of actions after block committed.
+    pub fn update_topology(&mut self, block: &VersionedCommittedBlock, new_peers: HashSet<PeerId>) {
+        self.lift_up_peers(
+            &block
+                .signatures()
+                .map(|s| s.public_key())
+                .cloned()
+                .collect::<Vec<PublicKey>>(),
+        );
+        self.rotate_set_a();
+        self.update_peer_list(new_peers);
+    }
+
+    /// Recreate topology for given block and view change index
+    pub fn recreate_topology(
+        block: &VersionedCommittedBlock,
+        view_change_index: usize,
+        new_peers: HashSet<PeerId>,
+    ) -> Self {
+        let mut topology = Topology::new(block.as_v1().header().committed_with_topology.clone());
+
+        topology.update_topology(block, new_peers);
+
+        // Rotate all once for every view_change
+        topology.rotate_all_n(view_change_index);
+
+        topology
+    }
+
+    /// Check if block's signatures meet requirements for given topology
+    ///
+    /// # Errors
+    /// - Not enough signatures
+    /// - Missing proxy tail signature
+    /// - Missing leader signature
+    pub fn verify_signatures<T>(
+        &self,
+        signatures: &mut SignaturesOf<T>,
+        hash: HashOf<T>,
+    ) -> Result<(), SignatureVerificationError> {
+        if !self.is_consensus_required() {
+            return Ok(());
+        }
+
+        let _ = signatures.retain_verified_by_hash(hash);
+
+        let votes_count = self
+            .filter_signatures_by_roles(
+                &[
+                    Role::ValidatingPeer,
+                    Role::Leader,
+                    Role::ProxyTail,
+                    Role::ObservingPeer,
+                ],
+                signatures.iter(),
+            )
+            .len();
+        let min_votes_for_commit = self.min_votes_for_commit();
+        if votes_count < min_votes_for_commit {
+            return Err(SignatureVerificationError::NotEnoughSignatures {
+                votes_count,
+                min_votes_for_commit,
+            });
+        }
+
+        if self
+            .filter_signatures_by_roles(&[Role::Leader], signatures.iter())
+            .is_empty()
+        {
+            return Err(SignatureVerificationError::LeaderMissing);
+        }
+
+        if self
+            .filter_signatures_by_roles(&[Role::ProxyTail], signatures.iter())
+            .is_empty()
+        {
```

Comment:
How do we tell that we have enough `ValidatinPeer` signatures? It's not obvious to me from the code.

Reply:
We check for total amount of signatures that we have at least 2f+1 signatures and then individually for proxy tail and leader signatures.

Reply:
Add as comments and I'll approve